### ISSUE TITLE
Support for React-Native

### ISSUE THREAD

**Issue by respectTheCode** [rollbar/rollbar.js#185]
We use rollbar for all of our web and native mobile projects but it doesn't work with react-native.

**Comment by antoinerousseau** [comment]
That would be great.

Sentry doc about it: https://docs.getsentry.com/hosted/clients/javascript/integrations/react-native/

Some discussions:
https://github.com/facebook/react-native/issues/1194
https://github.com/facebook/react-native/issues/5378

**Comment by dozoisch** [comment]
Same situation here!

**Comment by c-h-** [comment]
Same here!

**Comment by erickreutz** [comment]
We need this!

**Comment by antoinerousseau** [comment]
I have set up my RN app to call the Rollbar API when catching an exception, and it works great.
It looks like this (more or less):
```js
import {Platform} from 'react-native'
import StackTrace from 'stacktrace-js'

export const init = () => {
  const originalHandler = global.ErrorUtils.getGlobalHandler()

  const handler = (error) => {
    report(error, 'uncaught')
    if (originalHandler) {
      originalHandler(error)
    }
  }
  global.ErrorUtils.setGlobalHandler(handler)
}

export const report = (error, context) => {
  StackTrace.fromError(error, {offline: true})
  .then((stack) => {
    const frames = stack.map((frame) => {
      return {
        filename: frame.fileName,
        lineno: frame.lineNumber,
        colno: frame.columnNumber,
        method: frame.functionName,
      }
    })

    const params = {
      access_token: '<token>',
      data: {
        environment: (__DEV__ ? 'development' : 'production'),
        code_version: '1.0',
        platform: Platform.OS,
        framework: Platform.OS,
        context,
        body: {
          trace: {
            frames,
            exception: {
              class: error.name,
              message: error.message,
            },
          },
        },
      },
    }

    fetch('https://api.rollbar.com/api/1/item/', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(params),
    })
    .then((response) => {
      if (__DEV__ && response.status >= 400) {
        console.error('Rollbar response status is ' + response.status)
      }
    })
  })
}
```

However I couldn't manage to upload the source map to Rollbar.
```sh
react-native bundle --platform ios --dev false --entry-file index.ios.js --sourcemap-output main.jsbundle.map --bundle-output main.jsbundle
```
After multiple tries, I read this on the website:
> Some source map generation tools support an alternate representation of the map that combines multiple sub-maps into "sections" within the top level map. Each section can either be embedded source map or a reference to an external source map. Unfortunately, we don't yet support this source map format.

And my source map does indeed contain "sections".
Any thoughts?

**Comment by antoinerousseau** [comment]
Finally had to switch to Sentry. Their React Native plugin worked great for me, and they support source maps with sections.
https://github.com/antoinerousseau/tribeez/blob/master/src/mobile/report.js

**Comment by respectTheCode** [comment]
We switched to bugsnag after we couldn't get sentry to work.

**Comment by antoinerousseau** [comment]
Bugsnag looks neat indeed!
https://docs.bugsnag.com/platforms/react-native/

**Comment by brianr** [comment]
Hey everyone - just checking in to say we see this and we're planning to work on it soon. Will provide an update as soon as we have a timeline.

**Comment by tarr11** [comment]
Yeah, this is a showstopper for me.  Will probably switch soon if this isn't fixed.  Love rollbar otherwise though.

**Comment by skylarmb** [comment]
+1 ! We love it on web but would like to use it in react native as well

**Comment by rokob** [comment]
I am working on library support for react-native. Because of the unique environment, the plan is actually to have a different notifier library that deals with this use case directly. Much work needs to be done shuttling things across the js bridge into the native environments and this would not be well suited to live in the rollbar.js library. I will update this issue with further information as the work progresses.

**Comment by cdesch** [comment]
@rokob Can you share the library so we can help with it?

**Comment by stereodenis** [comment]
@brianr what status now?

**Comment by dmazin** [comment]
please friends I need this

**Comment by bmishkin** [comment]
we're considering moving to another exception product, bc rollbar doesn't have react native support =(

**Comment by symptomatik** [comment]
Would love this, please and thank you!!

**Comment by tlvenn** [comment]
@rokob any update on your progress ?

**Comment by maxs15** [comment]
Still doesn't support react-native ?
This is crazy ! :(

**Comment by r3nya** [comment]
Waiting for https://github.com/rollbar/rollbar-react-native :)

**Comment by zyskowsk** [comment]
@rokob excited for this to be done :) (also hi from everyone over here !)

**Comment by rokob** [comment]
Sorry for the delay on this, we made the decision to block this on some improvements to our Android notifier (namely rewriting the Java notifier and building a new Android one on top of, which was the last couple weeks: https://github.com/rollbar/rollbar-java/pull/66). That is now doneish, so I am anticipating getting this done quickly.

The plan is to have a new library: https://github.com/rollbar/rollbar-react-native which has rollbar.js, rollbar-ios, and rollbar-android as dependencies.

Getting rollbar.js and rollbar-android into a solid place to rely on them for react-native seemed like an opportunity to kill two birds so that was proritized over this issue. rollbar-ios could probably use some love too, but that will come after react-native.

Hey @zyskowsk, you guys using react-native? Bonjour to everyone over there, looks like things are going well.

**Comment by zyskowsk** [comment]
Everyone says bonjour back @rokob!  Yup, we released iOS/android apps built on react-native in september.  We are on bugsnag right now but will move over as soon as this is ready :)

**Comment by gregblass** [comment]
Just wanted to state that there are lots of people out there using Expo - which prevents us from using libraries that require native linking. If the rollbar integration could avoid using native linking, that would be ideal! Or at least have an option not to.

**Comment by rgangopadhya** [comment]
I use rollbar without linking:

```
import rollbar from 'rollbar/src/react-native/rollbar';
export const Rollbar = new rollbar({
  accessToken: ENV.ROLLBAR_TOKEN,
  captureUncaught: true,
  payload: {
     ...
  }
});
```

This works great except sourcemaps don't work -- as mentioned above, Expo generates sourcemaps with sections. Would be great if Rollbar supported those.

**Comment by rokob** [comment]
We are working on getting sectioned source maps working right now.

**Comment by rgangopadhya** [comment]
@rivkahstandig3636 @rokob I also notice that telemetry support, specifically `autoInstrument`, is missing here. Would be super helpful to be able to see recent activity before an error. Any plans to support this for react native?

**Comment by rokob** [comment]
Once we solidify what makes sense to instrument on mobile devices we intend to support telemetry. We could add manual events sooner than that, not sure on the timeline of that however.
